edge/pkg/edged/edged_pods.go

```diff
@@ -694,9 +721,74 @@ func (e *edged) makeEnvironmentVariables(pod *v1.Pod, container *v1.Container, p
 	for k, v := range tmpEnv {
 		result = append(result, kubecontainer.EnvVar{Name: k, Value: v})
 	}
+
+	// Append remaining service env vars.
+	for k, v := range serviceEnv {
+		// Accesses apiserver+Pods.
+		// So, the master may set service env vars, or kubelet may.  In case both are doing
+		// it, we skip the key from the kubelet-generated ones so we don't have duplicate
+		// env vars.
+		// TODO: remove this next line once all platforms use apiserver+Pods.
+		if _, present := tmpEnv[k]; !present {
+			result = append(result, kubecontainer.EnvVar{Name: k, Value: v})
+		}
+	}
 	return result, nil
 }
 
+var masterServices = sets.NewString("kubernetes")
+
+func (e *edged) getServiceEnvVarMap(ns string, enableServiceLinks bool) (map[string]string, error) {
+	var (
+		serviceMap = make(map[string]*v1.Service)
+		m          = make(map[string]string)
+	)
+
+	// Get all service resources from the master (via a cache),
+	// and populate them into service environment variables.
+	if e.serviceLister == nil {
+		// Kubelets without masters (e.g. plain GCE ContainerVM) don't set env vars.
+		return m, nil
+	}
+	services, err := e.serviceLister.List(labels.Everything())
+	if err != nil {
+		return m, fmt.Errorf("failed to list services when setting up env vars")
+	}
+
+	// project the services in namespace ns onto the master services
+	for i := range services {
+		service := services[i]
+		// ignore services where ClusterIP is "None" or empty
+		if !v1helper.IsServiceIPSet(&service) {
+			continue
+		}
+		serviceName := service.Name
+
+		// We always want to add environment variabled for master services
+		// from the master service namespace, even if enableServiceLinks is false.
+		// We also add environment variables for other services in the same
+		// namespace, if enableServiceLinks is true.
+		//todo masterservice's namespace hard-coded temporarily
+		if service.Namespace == "default" && masterServices.Has(serviceName) {
```

Comment:
Only for kubenetes default namespace?

Reply:
why? is it necessary?

Reply:
> Only for kubenetes default namespace? Should we consider kube-system namespace?

 The kubernetes service environment variable must be injected in the default namespace, and which will inject the service environment variable in the same namespace as the pod and can be controlled by enableServiceLinks field.

@daixiang0

Reply:
Get.